### PR TITLE
treewide: Disable automatic argparse argument shortening

### LINKED ISSUE #53495: RFC: treewide: python: argparse default configuration allows shortened command arguments
## Introduction

An issue has been identified in the argparse python library with its default configuration in which it enables automatic argument abbreviation (not user-set abbreviation as is common place in command line utilities). This causes bad script usage because commands can be added in the future which then silently and wholly without warning change the behaviour of what the script does.

### Problem description

This seems to be a whole zephyr-wide issue with any python code making use of argparse, this library decided that defaulting to allowing shortened command line arguments was a good idea, unlike traditional commands where you have a short and long argument e.g. `-h` and `--help`, this python library will, assuming you have a `--help` command accept this as `--h`, `--he`, `--hel` and `--help`. This is a pretty huge oversight as if a command exists, is wrongly called with a short argument and then a new command is added which uses the shortened argument, this will silently change the whole behaviour of the script, and indeed 2 whole test scripts for west flash commands start failing when this option is disabled and one starts failing because a new command has been added which the test script was wrongly using for the command previous.

Default description: https://docs.python.org/3/library/argparse.html#allow-abbrev

Example after disabling feature for one of the "west flash" runners:
```
tests/test_mdb.py FFF.......                                                                                                  [ 71%]

============================================================= FAILURES ==============================================================
____________________________________________________ test_flash_nsim[test_case0] ____________________________________________________

require = <MagicMock name='require' id='140064864486976'>, cc = <MagicMock name='popen_ignore_int' id='140064867421152'>
t = <MagicMock name='sleep' id='140064865197200'>, gcp = <MagicMock name='get_cld_pid' id='140064864890752'>
test_case = {'i': ['--cores=1', '--nsim=test_nsim.args'], 'o': ['mdb', '-nooptions', '-nogoifmain', '-toggle=include_local_symbols=1', '-nsim', '@/test/zephyr/boards/test-arch/test-board/support/test_nsim.args', ...]}
mdb_nsim = <function mdb.<locals>._factory at 0x7f6364cef490>

    @pytest.mark.parametrize('test_case', TEST_NSIM_FLASH_CASES)
    @patch('runners.mdb.get_cld_pid', return_value=(False, -1))
    @patch('time.sleep', return_value=None)
    @patch('runners.mdb.MdbNsimBinaryRunner.popen_ignore_int')
    @patch('runners.core.ZephyrBinaryRunner.require', side_effect=require_patch)
    def test_flash_nsim(require, cc, t, gcp, test_case, mdb_nsim):
>       mdb_nsim(test_case['i']).run('flash')

/tmp/aa/zephyr/scripts/west_commands/tests/test_mdb.py:157: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/tmp/aa/zephyr/scripts/west_commands/tests/test_mdb.py:126: in _factory
    arg_namespace = parser.parse_args(args)
/usr/lib/python3.10/argparse.py:1829: in parse_args
    self.error(msg % ' '.join(argv))
/usr/lib/python3.10/argparse.py:2587: in error
    self.exit(2, _('%(prog)s: error: %(message)s\n') % args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = ArgumentParser(prog='__main__.py', usage=None, description=None, formatter_class=<class 'argparse.HelpFormatter'>, conflict_handler='error', add_help=True)
status = 2, message = '__main__.py: error: unrecognized arguments: --nsim=test_nsim.args\n'

    def exit(self, status=0, message=None):
        if message:
            self._print_message(message, _sys.stderr)
>       _sys.exit(status)
E       SystemExit: 2

/usr/lib/python3.10/argparse.py:2574: SystemExit
------------------------------------------------------- Captured stderr call --------------------------------------------------------
usage: __main__.py [-h] [--elf-file FILE] [--hex-file FILE] [--bin-file FILE]
                   [--cores CORES] [--nsim_args NSIM_ARGS]
__main__.py: error: unrecognized arguments: --nsim=test_nsim.args

..
```

### Proposed change

A tree-wide change request for fixing this issue so that argparse is set to not allow it to generate its own shortened commands, when commands are added they can manually specify the shortened command to allow alongside the long argument.

## Detailed RFC

This requires setting `allow_abbrev=False` on all `ArgumentParser()` instantiations, and will require updating any script usage that does not use the full long argument name (or specified short argument name, if added).

### Dependencies

New python code which uses argparse should have `allow_abbrev=False` included and any invokations of scripts should use the proper argument names.

### Concerns and Unresolved Questions

This will only impact code in zephyr, it may also need propegating to other zephyr-hosted repositories and they should also be updated (if they have python scripts using argparse) to set `allow_abbrev=False` to prevent future issues with those scripts too.

User scripts for custom builds might also need fixing, this is unknown.

## Alternatives

Leave it as it is, which is a big risk with silent breakage, and possible security risk to CI tasks (i.e. user/customer production CI builds)

### PR DESCRIPTION
**Pull request by nordicjm**
Disables allowing the python argparse library from automatically shortening command line arguments, this prevents issues whereby a new command is added and code that wrongly uses the shortened command of an existing argument which is the same as the new command being added will silently change script behaviour.

Fixes #53495

- [X] Scripts updated
- [X] Script invocations updated
- [x] Release notes added
- [X] RFC treewide change email sent
- [X] Architecture working group discussed
- [X] pylint check added to ensure additional python code isn't added that uses argparse without this
- [x] Have pylint check run in CI

### CODE REVIEW

**Comment by nordicjm** [comment]
@stephanosio would it be possible to run a full CI on this (the weekly one) just to check as best as possible that it hasn't left anything broken?

**Comment by stephanosio** [comment]
> @stephanosio would it be possible to run a full CI on this (the weekly one) just to check as best as possible that it hasn't left anything broken?

One of the reasons that we run "full CI" on bi-weekly basis rather than daily or per PR is that it is very expensive to run.

Unless there is a very good reason as to why this PR needs to be treated specially, the full CI run should be run on schedule after it is merged -- if anything is broken, either this can be reverted or hotfixes can be provided.

**Comment by nordicjm** [comment]
> Unless there is a very good reason as to why this PR needs to be treated specially, the full CI run should be run on schedule after it is merged -- if anything is broken, either this can be reverted or hotfixes can be provided.

I think code in zephyr is OK, but what I'm concerned about is scripts in and/or being called by other repos e.g. HAL modules, can wait for a full run though.

**Comment by nordicjm** [comment]
Compliance issues are from existing code which is unchanged in this PR.

**Review by dcpleung — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

**Review by de-nordic — DISMISSED** [review]

**Comment by carlescufi** [comment]
Architecture WG:

- Requires an addition in the release notes
- Requires a pylint custom rule to verify this during compliance on PR
- No objections during the meeting

**Comment by nordicjm** [comment]
Added a pylint checker for this, but unsure of how to call it, according to the docs https://pylint.pycqa.org/en/latest/development_guide/how_tos/custom_checkers.html it needs to either be a full python package or in the PYTHON_PATH - so how would the python path be updated for the CI build to include the additional directory?
Also used 99 for the base group because their documentation just says it needs to be unique and not conflict with any of their base groups without listing what they are, so 99 seems like a safe bet.

**Review by mbolivar-nordic — DISMISSED** [review]

**Review by de-nordic — DISMISSED** [review]
migrate_mcmgr_kconfigs.py looks OK.

File: scripts/ci/check_compliance.py [hunk 1]
```diff
@@ -826,14 +830,23 @@ def run(self):
         if not py_files:
             return
 
-        pylintcmd = ["pylint", "--rcfile=" + pylintrc] + py_files
+        python_environment = os.environ.copy()
+        if "PYTHONPATH" in python_environment:
+            python_environment["PYTHONPATH"] = check_script_dir + ":" + \
+                                               python_environment["PYTHONPATH"]
+        else:
+            python_environment["PYTHONPATH"] = check_script_dir
```
Comment (fabiobaltieri) [inline scripts/ci/check_compliance.py:838]: You could always set and only handle the extend case:

```
python_environment["PYTHONPATH"] = check_script_dir 
if "PYTHONPATH" in os.environ:
  python_environment["PYTHONPATH"] += ":" + os.environ['PYTHONPATH"]
```

saves an "else", up to you really.

**Review by fabiobaltieri — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by nordicjm** [comment]
@fabiobaltieri can you merge this overriding the compliance check?

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri can you merge this overriding the compliance check?

Nope, looks like it's marked as required, I have no permissions to bypass it. @carlescufi can you do it?

**Review by stephanosio — APPROVED** [review]

**Pull request merged**